File: src/main/scala/org/cdp/rhasta/algorithm/sort/BubbleSort.scala
package org.cdp.rhasta.algorithm.sort

class BubbleSort {

  // 类java
  def bubbleSort1(arr: Array[Int]): Array[Int] = {
    for {
      i <- 0 until arr.length - 1
      j <- 0 until arr.length - i - 1
    } {
      if (arr(j) >= arr(j + 1)) {
        val temp = arr(j)
        arr(j) = arr(j + 1)
        arr(j + 1) = temp
      }
    }
    arr
  }

  // 纯scala
  import scala.annotation.tailrec
  def bubbleSort2(xt: List[Int]): List[Int] = {
    @tailrec
    def bubble(xs: List[Int],rest: List[Int], sorted: List[Int]): List[Int] = xs match {
      case x :: Nil =>
        if (rest.isEmpty) x :: sorted
        else bubble(rest, Nil, x :: sorted)
      case a :: b :: xs =>
        if (a > b) bubble(a :: xs, b :: rest, sorted)
        else bubble(b :: xs, a :: rest, sorted)
    }
    bubble(xt,Nil, Nil)
  }
}

object BubbleSort {
  val self = new BubbleSort

  def main(args: Array[String]): Unit = {
    val a = Array(5,3,19,2,1,10)

    val b = self.bubbleSort1(a)
    println(b.mkString(","))

    val c = self.bubbleSort2(a.toList)
    println(c)
  }
}

File: src/main/java/org/cdp/rhasta2/algorithm/sort/BubbleSort.java
package org.cdp.rhasta2.algorithm.sort;

import java.util.Arrays;

public class BubbleSort {
    public static <E extends Comparable<? super E>> void bubbleSort1(E[] comparable) {
        boolean changed = false;
        do {
            changed = false;
            for (int i = 0; i < comparable.length - 1; i++) {
                if (comparable[i].compareTo(comparable[i + 1]) > 0) {
                    E tmp = comparable[i];
                    comparable[i] = comparable[i + 1];
                    comparable[i + 1] = tmp;
                    changed = true;
                }
            }
        } while (changed);
    }

    public static void bubbleSort2(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            for (int j = 0; j < arr.length - 1; j++) {
                if (arr[j] > arr[j + 1]) {
                    int tmp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = tmp;
                }
            }
        }
    }

    public static void main(String[] args) {
        Integer[] a = {1,10,2,19,0,-1};
        bubbleSort1(a);
        for (Integer i: a) {
            System.out.print(i + " ");
        }
        System.out.println();

        int[] b = {1,10,2,19,0,-1};
        bubbleSort2(b);
        for (int i: b) {
            System.out.print(i + " ");
        }
    }
}
